### ISSUE TITLE
Arbitrary input validates against `Void` schema 

### ISSUE THREAD

**Issue by plcplc** [NorfairKing/autodocodec#34]
Considering the test example in `autodocodec-api-usage/test_resources/swagger-schema/either-void-bool.json`, running that through a json schema validator seems to accept arbitrary input for the `Left` case, whereas I would expect the set of valid inputs to be empty.

**either-void-bool.json, inlining definition "(Either Void Bool)":**

```json
{
  "properties": {
    "Left": {
      "additionalProperties": true
    },
    "Right": {
      "type": "boolean"
    }
  },
  "type": "object",
  "info": {
    "title": "",
    "version": ""
  },
  "paths": {},
  "swagger": "2.0"
}
```

Using https://www.jsonschemavalidator.net/, this appears to accept arbitrary input:
![image](https://user-images.githubusercontent.com/358550/216031939-45440ee8-1862-4ffa-8f7a-cb84ef370fbc.png)

I would have expected something similar to:

```
{
  "properties": {
    "Left": {
      "additionalProperties": true,
      "not": {
        "anyOf": [
          {"type": "string"},
          {"type": "number"},
          {"type": "object"},
          {"type": "null"},
          {"type": "array"},
          {"type": "boolean"}
        ]
      }
    },
    "Right": {
      "type": "boolean"
    }
  },
  "type": "object",
  "info": {
    "title": "",
    "version": ""
  },
  "paths": {},
  "swagger": "2.0"
}
```

![image](https://user-images.githubusercontent.com/358550/216032284-33719214-4ba5-47ef-b401-8bede9e817c3.png)

Which ought to reject any input.

I'd like to point out that I'm by no means an authority at interpreting openapi/jsonschema/..., but this behavior does seem wrong to me.

**Comment by plcplc** [comment]
Or even 
```json
{
  "properties": {
    "Left": {
      "additionalProperties": true,
      "not": {}
    },
    "Right": {
      "type": "boolean"
    }
  },
  "type": "object",
  "info": {
    "title": "",
    "version": ""
  },
  "paths": {},
  "swagger": "2.0"
}
```

**Comment by plcplc** [comment]
> Because bimapCodec allows arbitrary parse-failures, there's no way for the schema to be as strict as the codec.

That is true - but it also feels like a low aspiration for a dedicated `VoidCodec` constructor :-)

> I'm open to suggestions about how to fix this.

I _think_ that just producing the following json schema for a `VoidCodec` is a passable solution:

```json
{ "additionalProperties": true,
  "not": {}
}
```

Alternatively, the approach taken by @i-am-tom in https://github.com/NorfairKing/autodocodec/pull/32 which transitively censors void-containing branches from the schema is a good approach too I think (and one that results in a more informative schema).

**Comment by i-am-tom** [comment]
@NorfairKing If it doesn't seem useful, then it's worth pointing out that `Either Void` is only the easiest degenerate example. Consider the following API type, or anything with a similar [Trees That Grow](https://www.microsoft.com/en-us/research/uploads/prod/2016/11/trees-that-grow.pdf) feel to it:

```haskell
data IdentifierQuery t
  = ByKey (Key t)
  | ByIndex (Index t)
```

This hypothetical type allows us to query different types of data store with a common interface, but by using `Void` as a `Key` or `Index` type, we can remove parts of the interface that don't make sense, and do so with full support from the exhaustivity checker. For example, we can imagine:

```haskell
class CanQuery t where
  type Key t :: Type
  type Index t :: Type

instance CanQuery Array where
  type Key Array = Void
  type Index Array = Int

instance CanQuery (Map k) where
  type Key (Map k) = k
  type Index (Map k) = Void

instance CanQuery Set where
  type Key Set = Void
  type Index Set = Void
```

We can query an array by index but not field, we can query an object by field but not index, and we can query a set by neither. With that in mind, I would like to be able to write:

```haskell
instance (HasCodec (Key t), HasCodec (Index t)) => HasCodec (Query t)
```

... and have that instance specialise to the various different types throughout my schema, as I can with e.g. `ToJSON`/`FromJSON` instances. This is why `HasCodec Void` is useful.

---

Pushing a bit further, though, we can imagine another API type:

```haskell
data UpdateName t
  = UpdateName
       { newName :: String
       , target :: IdentifierQuery t
       }

instance (HasCodec (Key t), HasCodec (Index t))
  => HasCodec (UpdateName t)
```

So, what does the schema look like for `UpdateName Set`? This is why "bubbling up" `Void` (and collapsing `Void` sides of a sum) is useful: there are whole branches of complex structures that end up not existing.

> The `HasCodec Void` instance was probably a mistake in the sense that it's not useful

This made me a bit sad to read, so I'm sorry about the wall of text. The problem is that we use this library a lot (and like doing so), but the schema produced is not something we could currently release to the public. Our codebase uses this trick a _lot_ to try to maintain consistency where possible across different implementations of an interface.

That said, maybe this is a reasonably unique problem to us, and if TTG-style types are a niche that isn't worth the burden, then that's our problem, not `autodocodec`'s :)

**Comment by plcplc** [comment]
@NorfairKing: Just want to let you know that while I do want to work on this, we're somewhat short on time for API-codecs related stuff at the moment. I'll revisit if/when that changes in the future.
